keps/sig-node/3008-qos-class-resources/README.md

```diff
@@ -0,0 +1,1899 @@
+    field (in PodSpec) in which case it is regarded as a default for all
+    containers within that pod
+  - resource names across pod and container level QoS-class resource names
+    must be unique - i.e. the runtime cannot register a pod-level and
+    container-level resource with the same name
+- runtime advertises available QoS resources to kubelet
+  - QoS resources are dynamic i.e. can change during the lifetime of the node
+  - CRI [RuntimeStatus](#runtimestatus) message is used for carrying the information
+- kubelet
+  - updates NodeStatus based on the available QoS-class resources advertised by
+    the runtimea
+  - relays QoS-class resource requests from PodSpec to the runtime at pod
+    startup via [ContainerConfig](#containerconfig) and
+    [PodSandboxConfig](#podsandboxconfig) messages
+  - implements admission handler that validates the availability of QoS-class resources
+  - pod eviction as a possible [future improvement](#kubelet-initiated-pod-eviction)
+- scheduler
+  - simple node filtering based on the availability of QoS-class resources in
+    the first implementation phase (with possible
+    [future improvements](#scheduler-improvements))
+  - pod priority based eviction is in effect
+- quota, limits and defaults are [future work](#future-work) items
+
+
+### CRI API
+
+The following additions to the CRI protocol are suggested.
+
+#### ContainerConfig
+
+The `ContainerConfig` message will be supplemented with new `class_resources`
+field, providing per-container setting for QoS-class resources. This will be
+used in `CreateContainerRequest` to communicate the container-level QoS-class
+resource assignments to the runtime.
+
+```diff
+ message ContainerConfig {
+
+ ...
+     // Configuration specific to Linux containers.
+     LinuxContainerConfig linux = 15;
+     // Configuration specific to Windows containers.
+     WindowsContainerConfig windows = 16;
++
++    // Configuration of QoS resources.
++    ContainerQoSResources qos_resources = 17;
+ }
+
++// ContainerQoSResources specifies the configuration of QoS resources of a
++// container.
++message ContainerQoSResources {
++    // QoS resources the container will be assigned to.
++    // Key-value pairs where key is name of the QoS resource and value is the
++    // name of the class.
++    map<string, string> classes = 1;
++}
+```
+
+#### UpdateContainerResourcesRequest
+
+Similar to `CreateContainerRequest`, the `UpdateContainerResourcesRequest`
+message will extended to allow updating of QoS-class resource configuration of
+a running container.  Depending on runtime-level support of a particular
+resource (and possibly the type of resource) UpdateContainerResourcesRequest
+might fail. Resource discovery (see [Runtime status](#runtime-status)) has
+the capability to distinguish immutable resource types.
+
+Note that neither of the existing QoS-class resource types (RDT or blockio)
+support updates because of runtime limitations, yet.
+
```

Comment:
a note here and in the CRI message should be that UpdateContainerResources requests need to be atomic and fail if an aspect isn't applied. Thus, the container runtime should screen for the possibility of updating the resources before it calls the oci runtime to update the other resources

Reply:
Correct,. The wording here was somewhat confusing. I now updated the text, making your point clear.

Reply:
I also updated the comment in the CRI message